File: api-core/src/main/java/com/laynemobile/api/processor/ProcessorBuilder.java
/*
 * Copyright 2016 Layne Mobile, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.laynemobile.api.processor;

import com.laynemobile.api.Builder;
import com.laynemobile.api.processor.Processor.Checker;
import com.laynemobile.api.processor.Processor.Extension;
import com.laynemobile.api.processor.Processor.Interceptor;
import com.laynemobile.api.processor.Processor.Modifier;
import com.laynemobile.api.types.TypeBuilder;

import java.util.ArrayList;
import java.util.List;

public abstract class ProcessorBuilder<T, P, H> implements Builder<Processor<T, P>> {
    private final List<ProcessorHandler<T, P, ? extends H>> handlers = new ArrayList<>();

    private ProcessorBuilder() {}

    public final ProcessorBuilder<T, P, H> add(ProcessorHandler<T, P, ? extends H> handler) {
        handlers.add(handler);
        return this;
    }

    protected abstract Processor<T, P> processor(H h);

    @Override public final Processor<T, P> build() {
        ImmutableInterceptProcessor.Builder<T, P> builder
                = ImmutableInterceptProcessor.builder();

        final H h = buildType();
        for (ProcessorHandler<T, P, ? extends H> handler : handlers) {
            Extension<T, P> extension = extension(handler, h);
            if (extension instanceof Checker) {
                builder.addCheckers((Checker<T, P>) extension);
            } else if (extension instanceof Modifier) {
                builder.addModifiers((Modifier<T, P>) extension);
            } else if (extension instanceof Interceptor) {
                builder.addInterceptors((Interceptor<T, P>) extension);
            } else {
                throw new IllegalArgumentException("unknown Extension type: " + extension);
            }
        }

        return builder.setProcessor(processor(h))
                .build();
    }

    private H buildType() {
        final TypeBuilder<H> typeBuilder = new TypeBuilder<>();
        for (ProcessorHandler<T, P, ? extends H> handler : handlers) {
            typeBuilder.module(handler.typeHandler());
        }
        return typeBuilder.build();
    }

    @SuppressWarnings("unchecked")
    private <S extends H> Extension<T, P> extension(ProcessorHandler<T, P, S> handler, H h) {
        return handler.extension((S) h);
    }
}

File: api/src/main/kotlin/com/laynemobile/api/processor/ProcessorBuilder.kt
/*
 * Copyright 2016 Layne Mobile, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.laynemobile.api.processor

import com.laynemobile.api.Builder
import com.laynemobile.api.util.copy
import java.util.*

open class ProcessorBuilder<T : Any, R : Any?>
internal constructor() : Builder<Processor<T, R>> {

    private var _processor: ((T) -> R)? = null
    private var _errorHandlerProcessor: ErrorHandlerProcessor<T, R>? = null
    private val _extensions: MutableList<Extension<T, R>> = ArrayList()

    protected val processor: ErrorHandlerProcessor<T, R>
        get() = _errorHandlerProcessor!!
    protected val extensions: List<Extension<T, R>>
        get() = _extensions.copy()

    open fun withProcessor(function: (T) -> R) {
        _processor = function
    }

    fun withProcessor(processor: ErrorHandlerProcessor<T, R>) {
        _errorHandlerProcessor = processor
    }

    fun withErrorHandler(errorHandler: ErrorHandler<R>) {
        val p = _errorHandlerProcessor ?: _processor
        _errorHandlerProcessor = p?.let { it.withErrorHandler(errorHandler) }
    }

    fun withErrorHandler(errorHandler: (Throwable) -> R) {
        withErrorHandler(errorHandler.toErrorHandler())
    }

    fun extend(extension: Extension<T, R>) {
        _extensions += extension
    }

    inline fun extend(block: () -> Extension<T, R>) {
        extend(block())
    }

    fun extend(vararg extensions: Extension<T, R>) {
        _extensions += extensions
    }

    fun check(checker: Extension.Checker<T>) {
        extend(checker)
    }

    fun check(checker: (T) -> Unit) = extend {
        object : Extension.Checker<T>() {
            override fun check(t: T): Unit = checker(t)
        }
    }

    fun modify(modifier: Extension.Modifier<T, R>) {
        extend(modifier)
    }

    fun modify(modifier: (T, R) -> R) = extend {
        object : Extension.Modifier<T, R>() {
            override fun modify(t: T, r: R): R = modifier(t, r)
        }
    }

    fun intercept(interceptor: Extension.Interceptor<T, R>) {
        extend(interceptor)
    }

    fun intercept(interceptor: (Extension.Interceptor.Chain<T, R>) -> R) = extend {
        object : Extension.Interceptor<T, R>() {
            override fun intercept(chain: Chain<T, R>): R = interceptor(chain)
        }
    }

    override fun build(): Processor<T, R> = InterceptProcessor(
            processor = processor,
            extensions = extensions
    )
}

fun <T : Any, R : Any?> buildProcessor(init: ProcessorBuilder<T, R>.() -> Unit): Processor<T, R> {
    val builder: ProcessorBuilder<T, R> = ProcessorBuilder()
    builder.init()
    return builder.build()
}
